src/overload_hessian.jl

```diff
@@ -1,128 +1,99 @@
-## 1-to-1
-for fn in ops_1_to_1_s
-    @eval Base.$fn(t::HessianTracer) = promote_order(t)
-end
-for fn in ops_1_to_1_f
-    @eval Base.$fn(t::HessianTracer) = t
-end
-
-for fn in union(ops_1_to_1_z, ops_1_to_1_const)
-    @eval Base.$fn(::T) where {T<:HessianTracer} = empty(T)
-end
-
-## 2-to-1
-# Including second-order only
-for fn in ops_2_to_1_ssc
-    @eval function Base.$fn(a::HessianTracer, b::HessianTracer)
-        a = promote_order(a)
-        b = promote_order(b)
-        return distributive_merge(a, b)
+# TODO REVIEW: this keeps popping up so often that think it's worth an abstraction.  
+# It was previously called "promote_order".
+# The overly descriptive name is TEMPORARY.
+function tracer_from_hessian_OR_outer_product_OR_of_gradient(
+    t::T
+) where {T<:GlobalHessianTracer}
+    hessian_out = deepcopy(t.hessian)
+    hessian_out = outer_product_or!(hessian_out, t.gradient, t.gradient)
+    return T(t.gradient, hessian_out)
+end
+
+# ## 1-to-1
+for fn in ops_1_to_1
+    @eval function Base.$fn(t::T) where {T<:GlobalHessianTracer}
+        if is_seconder_zero_global($fn)
+            if is_firstder_zero_global($fn)
+                return empty(T)
+            else
+                return t
+            end
+        else
+            return tracer_from_hessian_OR_outer_product_OR_of_gradient(t)
+        end
     end
-    @eval Base.$fn(t::HessianTracer, ::Number) = promote_order(t)
-    @eval Base.$fn(::Number, t::HessianTracer) = promote_order(t)
 end
 
-for fn in ops_2_to_1_ssz
-    @eval function Base.$fn(a::HessianTracer, b::HessianTracer)
-        a = promote_order(a)
-        b = promote_order(b)
-        return additive_merge(a, b)
+# ## 2-to-1
+for fn in ops_2_to_1
+    @eval function Base.$fn(a::T, b::T) where {G,H,T<:GlobalHessianTracer{G,H}}
+        gradient_out, hessian_out = if !is_firstder_arg1_zero_global($fn)
+            if !is_firstder_arg2_zero_global($fn)
+                (a.gradient ∨ b.gradient, a.hessian ∨ b.hessian)
+            else
+                (a.gradient, deepcopy(a.hessian))
+            end
+        else # is_firstder_arg1_zero_global
+            if !is_firstder_arg2_zero_global($fn)
+                (b.gradient, deepcopy(b.hessian))
+            else
+                (empty_sparse_vector(G), empty_sparse_matrix(H))
+            end
+        end
```

Comment:
is there a reason to have a first batch of `if / else` and then do in-place modifications? I feel like this makes things more obscure

Reply:
Let's look at the following equation (screenshot since GitHub fails to render it):
<img width="511" alt="image" src="https://github.com/adrhill/SparseConnectivityTracer.jl/assets/20258504/5cc3fa20-2da9-4bbe-96a6-50d60a07e8ec">

The "first batch" computes the union of $\nabla^2\alpha$ and $\nabla^2\beta$.
The second block then updates that allocated Hessian with the four outer-ORs that follow.

Reply:
I wonder if there is a way to cleverly represent the four outer-ORs together

but as you said above, let's stick to the theory first and optimize later

Reply:
As a reference for later:

Since for sets, the outer-ORs correspond to nested loops creating index-tuples `(i, j)`, it might be worth merging $\mathbb{1}[\nabla\alpha] \vee \mathbb{1}[\nabla\alpha]^T$ with $\mathbb{1}[\nabla\alpha] \vee \mathbb{1}[\nabla\beta]^T$ 
to $\mathbb{1}[\nabla\alpha] \vee (\mathbb{1}[\nabla\alpha] \vee \mathbb{1}[\nabla\beta])^T$ when possible.

Reply:
And now that I'm thinking about it, terms 5 and 6 should always correspond to
$(\mathbb{1}[\nabla\alpha] \vee \mathbb{1}[\nabla\beta]) \vee (\mathbb{1}[\nabla\alpha] \vee \mathbb{1}[\nabla\beta])^T$

Reply:
Alternatively, the matrices in term 5 and 6 can be created simultaneously, since they create index tuples `(i , j)` and `(j, i)` respectively.

Reply:
The terms 5 and 6 must definitely be handled together, since they are associated to the same partial derivative

I just wrote them separately in the math because of formatting